### ISSUE TITLE
Variables cannot have a `metadata: dict` attribute, unlike parameters

### ISSUE THREAD

**Issue by nikhilwoodruff** [openfisca/openfisca-core#1071]
Hi there!

I really enjoy OpenFisca, but I recently encountered an issue.

### Here is what I did:

Added a `metadata: dict` attribute to a Variable class.

### Here is what I expected to happen:

The attribute to be silently accepted.

### Here is what actually happened:

A `ValueError` was thrown, specifying that `metadata` is not in the allowed attribute list.

## Context

I identify more as a:

- Developer _(I create tools that use the existing OpenFisca code)_.

**Comment by maukoquiroga** [comment]
By the way an extract of https://github.com/openfisca/openfisca-core/pull/1034#issue-718117616 I think could be pertinent here:

> (...) This PR introduces the notion of `schemas`. But, contrary to regular `schemas` used to validate, serialise, and deserialise data from/to models, these schemas are purely type-safe data objects, meant for type-checks, without impact at runtime.
> 
> #### Extended note on data-schemas
> 
> What if, instead of just using `schemas` as typed-dicts for type-checks, we used them at runtime to compulsorily:
> 
> a. check types at runtime?
> b. validate data at runtime?
> c. serialise/deserialise data —as suggested in #1071 ?
> d. enforce contracts, or domain logic —both in terms of data ("has to be >0") and representation ("tojson()")?
> 
> ##### For
> 
> - Explicit declaration and documentation of the two most important data models of OpenFisca, today dynamic and implicit: `Formula` and `Test`.
> - Explicit declaration and documentation of interface contracts and data transmutation —this is already, partially, done with `OpenAPI`
> - Complete separation of actual model and representational logic —extracting json logic from TaxBenefitSystem, Holder, and so on!
> - Composability: once all data/logic are split, and data/representation through explicit schemas, then OpenFisca would become fully-modular. That is, users can use drop-in replacements tailored to their needs, as long as those replacements respect the interface contracts: data in messagepack, models in rust, a calculation engine nodejs, etc.
> 
> ##### Against
> 
> **Performance**: whereas declarative data transmutation can even increase performance for individual situations —WebAPI—, a naive or out-of-the-box implementation (or anything with a complexity of at least `O(n)`) will certainly have a very penalising impact on performance for large population simulations.
> 
> This due to several reasons:
> 
> - Economists already validate data _before_ running a simulation — @benjello can you confirm?
> - `numpy` already provides data type-casting, with configurable levels of safety, and C-optimised
> - Other that I cannot think of right now 🤣 
> 
> Possible workarounds:
> 
> - make it opt-in with a flag —for example an optional WebAPI flag `--safe` or `--strict`
> - implement validations, serialisation, etc., vectorially from the outset —literally extending the numpy C-API and vendoring C-extensions, pre-compiled or not

**Comment by benjello** [comment]
@maukoquiroga : yes we try to validate data values before injection. I also confirm that the performance of OpenFisca relies on vector computation so if we add checks that are not vectorizable it puts a huge penalty on performance.
